packages/react-async/src/hooks.ts

```diff
@@ -1,13 +1,121 @@
+  ...args: IfAllOptionalKeys<
+    PrefetchOptions,
+    [AsyncComponentType<any, any, any, PrefetchOptions, any>, PrefetchOptions?],
+    [AsyncComponentType<any, any, any, PrefetchOptions, any>, PrefetchOptions]
+  >
+): ReturnType<typeof args[0]['usePrefetch']> {
+  const [asyncComponent, options = {}] = args;
+  return (asyncComponent.usePrefetch as any)(options);
+}
+
+export function useKeepFresh<KeepFreshOptions extends object>(
+  ...args: IfAllOptionalKeys<
+    KeepFreshOptions,
+    [
+      AsyncComponentType<any, any, any, any, KeepFreshOptions>,
+      KeepFreshOptions?
+    ],
+    [AsyncComponentType<any, any, any, any, KeepFreshOptions>, KeepFreshOptions]
+  >
+): ReturnType<typeof args[0]['useKeepFresh']> {
+  const [asyncComponent, options = {}] = args;
+  return (asyncComponent.useKeepFresh as any)(options);
+}
+
+interface Options {
+  immediate?: boolean;
+  assets?: AssetTiming;
+  styles?: AssetTiming;
+  scripts?: AssetTiming;
+}
+
+export function useAsync<T>(
+  resolver: Resolver<T>,
+  {assets, scripts = assets, styles = assets, immediate = true}: Options = {},
+) {
+  const [value, setValue] = useState<T | Error | null>(
+    () =>
+      immediate || typeof window !== 'undefined' ? resolver.resolved : null,
+  );
+
+  const mounted = useMountedRef();
+
+  const load = useCallback(
+    async (): Promise<T | Error> => {
+      if (value != null) {
+        return value;
+      }
+
+      try {
+        const resolved = await resolver.resolve();
+
+        if (mounted.current) {
+          // It's important to use the function form of setValue here.
+          // Resolved is going to be a function in most cases, since it's
+          // a React component. If you do not set it using the function form,
+          // React treats the component as the function that returns state,
+          // so it sets state with the result of manually calling the component
+          // (so, usually JSX).
+          setValue(() => resolved);
+        }
+
+        return resolved;
+      } catch (error) {
+        if (mounted.current) {
+          setValue(error);
+        }
+
+        return error;
+      }
+    },
+    [mounted, resolver, value],
+  );
+
+  const {id} = resolver;
+
+  useAsyncAsset(id, {scripts, styles});
+
+  return value instanceof Error
+    ? {id, resolved: null, error: value, loading: false, load}
+    : {
+        id,
+        resolved: value,
+        error: null,
+        loading: value == null,
+        load,
+      };
+}
 
-export function useAsyncAsset(id?: string) {
-  const assets = useContext(AsyncAssetContext);
+export function useAsyncAsset(
+  id?: string,
+  {scripts, styles}: {styles?: AssetTiming; scripts?: AssetTiming} = {},
+) {
+  const async = useContext(AsyncAssetContext);
 
   useServerEffect(() => {
-    if (assets && id) {
-      assets.markAsUsed(id);
+    if (async && id) {
+      async.markAsUsed(id, {scripts, styles});
     }
-  });
+  }, async ? async.effect : undefined);
```

Comment:
🤔 this seems fishy, should we add our own throw for cases where this changes length so that it's a bit clearer if it gets hit?

Reply:
not sure what you mean by changes length here, the second argument to this function is an effect, not an array.